File: backend/app/engine/fog.ts
import type { GameState, VisibilityState } from './state.js'
import { UNIT_STATS, BASE_STATS } from './state.js'
import { hexesInRadius } from './hex.js'

/**
 * Recompute fog of war for all players using hex radius.
 */
export function computeFog(state: GameState): void {
  for (let slot = 0; slot < state.players.length; slot++) {
    const vis = state.visibility[slot]

    // Downgrade current 'visible' to 'previously_seen'
    for (let y = 0; y < state.mapConfig.height; y++) {
      for (let x = 0; x < state.mapConfig.width; x++) {
        if (vis[y][x] === 'visible') {
          vis[y][x] = 'previously_seen'
        }
      }
    }

    // Mark tiles visible from units
    for (const unit of state.units.values()) {
      if (unit.playerSlot !== slot) continue
      const vision = UNIT_STATS[unit.unitType].vision
      markVisible(vis, unit.x, unit.y, vision, state.mapConfig.width, state.mapConfig.height)
    }

    // Mark tiles visible from buildings
    for (const building of state.buildings.values()) {
      if (building.playerSlot !== slot) continue
      markVisible(
        vis,
        building.x,
        building.y,
        BASE_STATS.vision,
        state.mapConfig.width,
        state.mapConfig.height
      )
    }
  }
}

function markVisible(
  vis: VisibilityState[][],
  cx: number,
  cy: number,
  radius: number,
  mapWidth: number,
  mapHeight: number
): void {
  for (const pos of hexesInRadius({ x: cx, y: cy }, radius)) {
    if (pos.x >= 0 && pos.x < mapWidth && pos.y >= 0 && pos.y < mapHeight) {
      vis[pos.y][pos.x] = 'visible'
    }
  }
}

File: game/crates/bf_engine/src/fog.rs
use bf_types::*;
use crate::GameState;

/// Recompute fog of war for all players.
pub fn compute_fog(state: &mut GameState) {
    let num_players = state.players.len();
    let w = state.map_width as usize;
    let h = state.map_height as usize;

    // Ensure visibility arrays exist for all players
    while state.visibility.len() < num_players {
        state.visibility.push(
            vec![vec![VisibilityState::Unseen; w]; h]
        );
    }

    for slot_idx in 0..num_players {
        let slot = state.players[slot_idx].slot;

        // Step 1: Downgrade all 'visible' to 'previously_seen'
        for row in state.visibility[slot_idx].iter_mut() {
            for cell in row.iter_mut() {
                if *cell == VisibilityState::Visible {
                    *cell = VisibilityState::PreviouslySeen;
                }
            }
        }

        // Collect positions to mark visible
        let mut visible_positions: Vec<(Position, u32)> = Vec::new();

        // Step 2: Units grant visibility
        for unit in &state.units {
            if unit.player_slot == slot {
                let stats = unit_stats(unit.unit_type);
                visible_positions.push((Position::new(unit.x, unit.y), stats.vision));
            }
        }

        // Step 3: Buildings grant visibility
        for building in &state.buildings {
            if building.player_slot == slot {
                visible_positions.push((Position::new(building.x, building.y), BASE_VISION));
            }
        }

        // Mark visible hexes
        for (center, radius) in visible_positions {
            for pos in hexes_in_radius(center, radius) {
                if pos.x >= 0 && pos.y >= 0 && (pos.x as usize) < w && (pos.y as usize) < h {
                    state.visibility[slot_idx][pos.y as usize][pos.x as usize] = VisibilityState::Visible;
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{GameState, Unit, PlayerState};

    #[test]
    fn test_fog_marks_visible_around_unit() {
        let mut state = GameState {
            phase: MatchPhase::Active,
            tick: 0,
            map_width: 10,
            map_height: 10,
            terrain: vec![vec![Tile { tile_type: TileType::Grass, elevation: 1 }; 10]; 10],
            units: vec![Unit {
                id: "u1".to_string(),
                player_slot: 0,
                unit_type: UnitType::Scout,
                x: 5,
                y: 5,
                hp: 40,
                max_hp: 40,
                status: UnitStatus::Idle,
                path: vec![],
                target_id: None,
                cargo: 0,
            }],
            buildings: vec![],
            resource_nodes: vec![],
            players: vec![PlayerState { slot: 0, energy: 200, base_id: String::new() }],
            visibility: vec![],
            command_queue: vec![],
            winner_slot: None,
            snapshot_units: vec![],
            snapshot_buildings: vec![],
        };

        compute_fog(&mut state);

        // Scout has vision 6 — center should be visible
        assert_eq!(state.visibility[0][5][5], VisibilityState::Visible);
        // Far corner should still be unseen
        assert_eq!(state.visibility[0][0][0], VisibilityState::Unseen);
    }
}
